### PR TITLE
feat: Stream dual mode

### PR DESCRIPTION
**Pull request by BenFidge**
Fixes #2621

- Implements `IConnectorStreamModeSupport` interface

### CODE REVIEW

File: CluedIn.Connector.SqlServer.sln [hunk 1]
```diff
@@ -52,6 +52,12 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Connector.SqlServer", "src\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Connector.SqlServer.Unit.Tests", "test\unit\Connector.SqlServer.Test\Connector.SqlServer.Unit.Tests.csproj", "{5F2D36E4-2B4A-47EF-B4FC-703FFBB074E4}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Core", "..\CluedIn\Code\Core\Core.csproj", "{16946CAE-528E-4C2F-8E42-0DBAEFCE5557}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Core.Agent", "..\CluedIn\Code\Core.Agent\Core.Agent.csproj", "{68366B00-88D3-4063-91B2-ECEBD9FB2CAC}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Crawling", "..\CluedIn\Code\Crawling\Crawling.csproj", "{26C11789-BA1C-4C6A-BFD7-AEE3BBA885CE}"
+EndProject
 Global
```
Comment (Kieranties) [inline CluedIn.Connector.SqlServer.sln:61]: This is from debugging right? They should be removed

File: CluedIn.Connector.SqlServer.sln [hunk 2]
```diff
@@ -66,6 +72,18 @@ Global
...
+		{16946CAE-528E-4C2F-8E42-0DBAEFCE5557}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{16946CAE-528E-4C2F-8E42-0DBAEFCE5557}.Release|Any CPU.Build.0 = Release|Any CPU
+		{68366B00-88D3-4063-91B2-ECEBD9FB2CAC}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{68366B00-88D3-4063-91B2-ECEBD9FB2CAC}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{68366B00-88D3-4063-91B2-ECEBD9FB2CAC}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{68366B00-88D3-4063-91B2-ECEBD9FB2CAC}.Release|Any CPU.Build.0 = Release|Any CPU
+		{26C11789-BA1C-4C6A-BFD7-AEE3BBA885CE}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{26C11789-BA1C-4C6A-BFD7-AEE3BBA885CE}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{26C11789-BA1C-4C6A-BFD7-AEE3BBA885CE}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{26C11789-BA1C-4C6A-BFD7-AEE3BBA885CE}.Release|Any CPU.Build.0 = Release|Any CPU
```
Comment (Kieranties) [inline CluedIn.Connector.SqlServer.sln:86]: Same again - debug stuff to be removed.

File: Packages.props [hunk 3]
```diff
@@ -38,9 +38,9 @@
     -->
     <PackageReference Update="Castle.Windsor" Version="5.0.1" />
     <PackageReference Update="ComponentHost" Version="$(_ComponentHost)" />
-    <PackageReference Update="CluedIn.Core" Version="$(_CluedIn)" />
+    <!--<PackageReference Update="CluedIn.Core" Version="$(_CluedIn)" />
     <PackageReference Update="CluedIn.Core.Agent" Version="$(_CluedIn)" />
-    <PackageReference Update="CluedIn.Crawling" Version="$(_CluedIn)" />
+    <PackageReference Update="CluedIn.Crawling" Version="$(_CluedIn)" />-->
```
Comment (Kieranties) [inline Packages.props]: This needs to be reverted.

File: src/Connector.SqlServer/Connector.SqlServer.csproj [hunk 4]
```diff
@@ -12,18 +12,18 @@
...
   </ItemGroup>
 
   <!-- Disable to above cluedin assets and enable the below for local debugging  -->
-  <!--<ItemGroup>    
+  <ItemGroup>    
     <ProjectReference Include="..\..\..\Cluedin\Code\Core\Core.csproj" />
     <ProjectReference Include="..\..\..\Cluedin\Code\Core.Agent\Core.Agent.csproj" />
     <ProjectReference Include="..\..\..\Cluedin\Code\Crawling\Crawling.csproj" />
-  </ItemGroup>-->
+  </ItemGroup>
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector.SqlServer.csproj]: This needs to be reverted.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 5]
```diff
@@ -187,25 +212,35 @@ async Task RenameTable(string currentName, string updatedName, string context)
...
+        public override Task StoreData(ExecutionContext executionContext, Guid providerDefinitionId, string containerName, IDictionary<string, object> data)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override Task StoreEdgeData(ExecutionContext executionContext, Guid providerDefinitionId, string containerName, string originEntityCode, IEnumerable<string> edges)
+        {
+            throw new NotImplementedException();
+        }
+
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:238]: Why are these not implemented?

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 6]
```diff
@@ -338,10 +373,23 @@ public override async Task<bool> VerifyConnection(ExecutionContext executionCont
...
         {
             try
             {
+                // If we are in Event Stream mode, append extra fields
+                if (StreamMode == StreamMode.EventStream)
+                {
+                    data.Add(TimestampFieldName, timestamp);
+                    data.Add(ChangeTypeFieldName, changeType);
+                    data.Add(CorrelationIdFieldName, correlationId);
+                }
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:391]: We shouldn't add directly to the data object as it's a reference we'll be modifying the same object passed by the caller.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 7]
```diff
@@ -338,10 +373,23 @@ public override async Task<bool> VerifyConnection(ExecutionContext executionCont
...
+
                 if (_bulkSupported)
                 {
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:394]: This implementation should stay in the `StoreData` call - we should just be calling that with the updated values.

File: src/Connector.SqlServer/Features/IBuildStoreDataFeature.cs [hunk 8]
```diff
@@ -13,6 +15,10 @@ public interface IBuildStoreDataFeature
             string containerName,
             IDictionary<string, object> data,
             IList<string> keys,
+            StreamMode mode,
+            string correlationId,
+            DateTimeOffset timestamp,
+            VersionChangeType changeType,
```
Comment (Kieranties) [inline src/Connector.SqlServer/Features/IBuildStoreDataFeature.cs:21]: Breaking changes - We're introducing new interfaces to support the new types - if the old interface supports the default implementation, then introduce a new interface to support the new mode.

**Review by Kieranties — CHANGES_REQUESTED** [review]
We've got to be _really_ careful with breaking changes, especially in the area of connectors and crawlers as these are 'hot spots' for the implementation team.

File: CluedIn.Connector.SqlServer.sln (hunk 1 above)
Comment (BenFidge) [reply CluedIn.Connector.SqlServer.sln:61]: Reviewed too early lol. Was on draft. These have been reverted

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs (hunk 5 above)
Comment (BenFidge) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:238]: Never needed in 3.2.5

Comment (BenFidge) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:238]: Will remove actually

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs (hunk 6 above)
Comment (BenFidge) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:391]: Fair one

File: src/Connector.SqlServer/Features/IBuildStoreDataFeature.cs (hunk 8 above)
Comment (BenFidge) [reply src/Connector.SqlServer/Features/IBuildStoreDataFeature.cs:21]: Arent we adding mess + pain just to support Semler overrides which will take 5 mins to upgrade to 3.2.5?

Comment (Kieranties) [reply src/Connector.SqlServer/Features/IBuildStoreDataFeature.cs]: There are other connectors already created that are not for semler.  Every customer will be broken by this change and forced to update their connectors as well as server.

Non-of that is a 'five' minute upgrade - you have to consider the impact on production rollout and testing.

By keeping things working in a non-breaking manner users can upgrade relevant parts when ready.

**Comment by Kieranties** [comment]
@BenFidge Thanks for the updates - this will continue to fail while we don't have a 3.2.5 branch - I'll add this pr to the 3.2.5 ticket

**Comment by BenFidge** [comment]
> @BenFidge Thanks for the updates - this will continue to fail while we don't have a 3.2.5 branch - I'll add this pr to the 3.2.5 ticket

I thought this _was_ for 3.2.5? Lets have a chat tomorrow about handling the override DLL's (Semler etc)

**Review by Kieranties — CHANGES_REQUESTED** [review]
@BenFidge I've re-targeted the PR for the 3.2.5 branch - please could you add some release notes?

File: Packages.props [hunk 9]
```diff
@@ -44,8 +44,6 @@
     <PackageReference Update="CluedIn.CrawlerIntegrationTesting" Version="$(_CluedIn)" />
     <PackageReference Update="CluedIn.DataStore" Version="$(_CluedIn)" />
     <PackageReference Update="RestSharp" Version="106.10.1" />
-    <PackageReference Update="Dapper" Version="2.0.35" />
-    <PackageReference Update="Dapper.SqlBuilder" Version="2.0.35" />
```
Comment (Kieranties) [inline Packages.props:48]: Where are we sourcing these from now?

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 10]
```diff
@@ -187,25 +213,35 @@ async Task RenameTable(string currentName, string updatedName, string context)
...
+                    tasks.Add(renameTable(current.Value.Name, updatedContainer.Tables[current.Key].Name, current.Key));
             }
 
             await Task.WhenAll(tasks);
 
         }
 
+        public override async Task StoreData(ExecutionContext executionContext, Guid providerDefinitionId, string containerName, IDictionary<string, object> data)
+        {
+            await StoreData(executionContext, providerDefinitionId, containerName, null, DateTimeOffset.Now, VersionChangeType.NotSet, data);
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:232]: No need for await/async just return directly

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 11]
```diff
@@ -187,25 +213,35 @@ async Task RenameTable(string currentName, string updatedName, string context)
...
+        }
+
+        public override async Task StoreEdgeData(ExecutionContext executionContext, Guid providerDefinitionId, string containerName, string originEntityCode, IEnumerable<string> edges)
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:235]: No need for await/async just return directly

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 12]
```diff
@@ -338,18 +374,32 @@ public override async Task<bool> VerifyConnection(ExecutionContext executionCont
             }
         }
 
-        public override async Task StoreData(ExecutionContext executionContext, Guid providerDefinitionId, string containerName, IDictionary<string, object> data)
+        public async Task<string> GetCorrelationId()
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:377]: No need for await/async just return Task.FromResult

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 13]
```diff
@@ -416,7 +470,7 @@ public override async Task StoreData(ExecutionContext executionContext, Guid pro
                     // see if we need to delete linked tables
                     // do look up of OriginEntityCode from current table data
                     var lookupOriginCodes = new List<string>();
-                    using (var connection = await _client.GetConnection(config.Authentication))
+                    await using (var connection = await _client.GetConnection(config.Authentication))
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/SqlServerConnector.cs:473]: Not sure why the await is needed here?

File: src/Connector.SqlServer/Connector/Container.cs [hunk 14]
```diff
@@ -2,24 +2,33 @@
...
-            new ConnectionDataType { Name =  "Code".SqlSanitize(), Type = VocabularyKeyDataType.Text }
+        private static readonly ConnectionDataType[] _codeEdgeColumns = {
+            new ConnectionDataType { Name = "OriginEntityCode", Type = VocabularyKeyDataType.Text },
+            new ConnectionDataType { Name =  "Code", Type = VocabularyKeyDataType.Text }
         };
 
-        public Container(string containerName)
+        private static readonly ConnectionDataType[] _eventStreamCodeEdgeColumns = {new ConnectionDataType {Name = "CorrelationId", Type = VocabularyKeyDataType.Text},};
+
+        public Container(string containerName, StreamMode mode)
```
Comment (Kieranties) [inline src/Connector.SqlServer/Connector/Container.cs:18]: Breaking change - Keep the original constructor, calling this new on with the default mode (Steam.Sync?)
Then mark the original constructor as obsolete.

**Review by Kieranties — CHANGES_REQUESTED** [review]
Few tweaks here

File: Packages.props (hunk 9 above)
Comment (BenFidge) [reply Packages.props:48]: We've never used them, think they were dregs left over from when we first started

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs (hunk 13 above)
Comment (BenFidge) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:473]: Resharper suggestion

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs (hunk 10 above)
Comment (BenFidge) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:232]: Sure? 
image

**Comment by Kieranties** [comment]
@BenFidge We have failing tests that need fixing up

**Review by Kieranties — APPROVED** [review]

**Pull request merged**